### PR TITLE
HDDS-7546. [Snapshot] Throw exception if the SnapshotDiff size is huge

### PR DESCRIPTION
**Pull request by hemantk-12**
## What changes were proposed in this pull request?

This change is to throw exception if estimated keys changed in snapdiff job is more than 10 million.

## What is the link to the Apache JIRA
[HDDS-7546](https://issues.apache.org/jira/browse/HDDS-7546)

## How was this patch tested?
Existing unit tests for now.

### CODE REVIEW

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/ManagedSstFileReader.java [hunk 1]
```diff
@@ -49,16 +49,37 @@ public class ManagedSstFileReader {
...
+      return estimatedTotalKeys;
+    }
+
+    long estimatedSize = 0;
+    synchronized (this) {
+      try (ManagedOptions options = new ManagedOptions()) {
+        for (String sstFile : sstFiles) {
+          SstFileReader fileReader = new SstFileReader(options);
+          fileReader.open(sstFile);
+          estimatedSize += fileReader.getTableProperties().getNumEntries();
```
Comment (swamirishi) [inline hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/ManagedSstFileReader.java:75]: estimated Size here should be just getNumEntries() don't add the two. https://github.com/facebook/rocksdb/blob/main/include/rocksdb/table_properties.h#:~:text=//%20the%20number%20of,operands%20in%20the%20table

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/ManagedSstFileReader.java [hunk 2]
```diff
@@ -49,16 +49,37 @@ public class ManagedSstFileReader {
...
+          estimatedSize += fileReader.getTableProperties().getNumDeletions();
+        }
+      }
+    }
+
+    estimatedTotalKeys = estimatedSize;
```
Comment (swamirishi) [inline hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/ManagedSstFileReader.java:81]: Shouldn't we have this inside the synchronized block? And the synchronized block should also having the same if (estimatedTotalKeys != -1) {
      return estimatedTotalKeys;
    } to avoid redundant operations.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 3]
```diff
@@ -842,6 +853,20 @@ private Set<String> getDeltaFiles(OmSnapshot fromSnapshot,
     return deltaFiles;
   }
 
+  private void validateEstimatedKeyChangesAreInLimits(
+      ManagedSstFileReader sstFileReader
+  ) throws RocksDBException, IOException {
+    if (sstFileReader.getEstimatedTotalKeys() >
+        maxAllowedKeyChangesForASnapDiff) {
+      throw new IOException(
```
Comment (swamirishi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:861]: Nit: We can do this as part of a different PR, should we have a separate exception for Snapshots?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java [hunk 4]
```diff
@@ -155,9 +158,8 @@ public class SnapshotDiffManager implements AutoCloseable {
    */
   private final String sstBackupDirForSnapDiffJobs;
 
-  private boolean isNativeRocksToolsLoaded;
-
-  private ManagedSSTDumpTool sstDumpTool;
+  private final boolean isNativeRocksToolsLoaded;
```
Comment (swamirishi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:161]: Nit: In hindsight we could have just gotten away with just an Optional.

**Review by swamirishi — CHANGES_REQUESTED** [review]

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/ManagedSstFileReader.java (hunk 2 above)
Comment (hemantk-12) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/ManagedSstFileReader.java:81]: 1. It makes sense to move `estimatedTotalKeys = estimatedSize;` this in synchronized block. Will update it.
2. Regarding having double check `if (estimatedTotalKeys != -1)` inside the synchronized block, I agree but [findbug failed](https://github.com/apache/ozone/actions/runs/4813599053/jobs/8571917550) once when I did something similar to create singleton object. Another reason is that it doesn't matter that much because result is used by same thread. I'll add the check and see if findbug works. If does I'll update in next revision otherwise leave as it is.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java (hunk 3 above)
Comment (hemantk-12) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:861]: Yes, I'll add a TODO.

May be we can do it as part of [HDDS-8202(https://issues.apache.org/jira/browse/HDDS-8202).

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java (hunk 4 above)
Comment (hemantk-12) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:161]: Yes, we can use `Optional` but it will require some changes in [addToObjectIdMap](https://github.com/apache/ozone/blob/master/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java#L741). I'll leave it as it is in this PR and will address in [PR-4617](https://github.com/apache/ozone/pull/4617) since I already did some refactoring in that.

File: hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/ManagedSstFileReader.java (hunk 1 above)
Comment (hemantk-12) [reply hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdb/util/ManagedSstFileReader.java:75]: Thanks, I'll update it.

**Review by prashantpogde — APPROVED** [review]
LGTM

**Comment by prashantpogde** [comment]
@hemantk-12 Can we resolve the conflicts ? I believe the changes overall look good and can be merged after successful CI run.

**Review by swamirishi — APPROVED** [review]
LGTM

**Review by swamirishi — APPROVED** [review]
LGTM

**Comment by prashantpogde** [comment]
Thanks @hemantk-12 for this patch.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java (hunk 4 above)
Comment (hemantk-12) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:162]: Changed it to `Optional` in https://github.com/apache/ozone/pull/4617/files#diff-46f82d5fbac5dc7aae1fc363811fb88e6bc6c89a2aefafddad90de2f1ff3cfbdR163

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java (hunk 3 above)
Comment (hemantk-12) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java:920]: Added a TODO: https://github.com/apache/ozone/pull/4617/files#diff-46f82d5fbac5dc7aae1fc363811fb88e6bc6c89a2aefafddad90de2f1ff3cfbdR944
